### ISSUE TITLE
New JSX runtime advice for package authors?

### ISSUE THREAD

**Issue by rafgraph** [facebook/react#20491]
I'm wondering if you have any advice around using the new JSX runtime in packages with `react` as a peer dependency?

For compatibility reasons it seems like packages shouldn’t use the JSX runtime (at least not yet) because then the package would only be compatible with versions of React that include the new JSX runtime (but in time this compatibility reason will fade).

If a package did use the new JSX runtime, I think it could only be used in the ESM and CJS builds. The UMD build would still need to use `createElement` as the UMD version of React doesn't include the JSX runtime.

Curious if you could shed some light on the plan for the JSX runtime with regards to external packages? Will the JSX runtime eventually be part of the React UMD build? Does tooling need to support using the JSX runtime in ESM and CJS builds, while always using `createElement` in the UMD build? Thanks.

I am ran into this issue when creating [Rollpkg](https://github.com/rafgraph/rollpkg), a zero config build tool to create packages with Rollup and TypeScript.

Thanks for all your work on React, I've really enjoyed using it the past several years!

**Comment by phanithinks** [comment]
is there any workaround to use jsx in UMD build ?

**Comment by idler8** [comment]
look it 
https://github.com/reactjs/reactjs.org/issues/5659

**Comment by iegik** [comment]
TL;DR; https://github.com/facebook/react/issues/20923#issuecomment-1526015094

Here is the JavaScript modules example:
```
<script type="module">
  import React from 'react'; // Import core React
  import ReactDOM from 'react-dom/client'; // Import ReactDOM client
  import { Button } from "button";

  const jsx = React.createElement

  ReactDOM.hydrateRoot(document.getElementById("xyz"), jsx(Button, { title: "Press me" }));
</script>
```

Still question rising in the air: how can we use `ReactDOM.hydrateRoot` without JSX in JavaScript modules? Raw JSX is not supported, it throws syntax error. So, the only option is use `React.createElement` and pass props as an object but [`React.createElement`](https://react.dev/reference/react/createElement) is deprecated (inside Legacy API's section)

So, what we gonna do? Does React will automatically detect component props from the DOM tree? How it may be possible?

Some conditions is complicated and there is no chance, that you can predict that some of the props are set in particular way!

**Comment by iegik** [comment]
Ok. In response [to myself](https://github.com/facebook/react/issues/20491#issuecomment-2855696155), I think, I figured out. Since [ESM is coming](https://github.com/facebook/react/issues/20923#issuecomment-1255233148), I think I'm in right way. And `React.createElement` is not deprecated, but React Team just encourage you to use mostly JSX syntax. Thats it.

So, it is fine to use `React.createElement` in places where React elements where "hydrated" (until new ways where found) but this probably will be a single place in your app, where client and server stick together, because all other peaces will be nested children. So, no worries.
